sql/sql_parse.cc

```diff
@@ -6146,7 +6146,11 @@ static TABLE *find_temporary_table_for_rename(THD *thd,
     }
   }
   if (!found)
+  {
+    if (thd->has_temporary_tables())
+      thd->open_temporary_table(table);
     res= thd->find_temporary_table(table, THD::TMP_TABLE_ANY);
+  }
   DBUG_RETURN(res);
```

Comment:
I think it's the wrong place to fix it. The rule is that there cannot be a `TMP_TABLE_SHARE` without a single `TABLE`. It's violated not here, it's violated in `THD::find_temporary_table`, so I believe it must be fixed there.